zebra-network/src/peer_set/initialize.rs

```diff
@@ -605,6 +610,7 @@ where
 
             if active_inbound_connections.update_count()
                 >= config.peerset_inbound_connection_limit()
+                || recent_inbound_connections.is_past_limit_or_add(addr.ip())
             {
                 // Too many open inbound connections or pending handshakes already.
```

Comment:
Great catch.

> Can we add a short delay after dropping the connection?

Yes.

> The change above limits the number of entries to MIN_PEER_RECONNECTION_DELAY / MIN_INBOUND_PEER_FAILED_CONNECTION_INTERVAL = 11,900

It may not limit the number of entries at all, it would currently drop the connection and continue with the loop if an attacker used the same IP repeatedly. Great catch.

> Do you think it would take a lot of CPU to create and prune twelve thousand entries?

Yeah, far more than warranted.

> Should we put a lower limit on the number of entries?

Good idea, yeah, there should be a constant lower limit.

Reply:
> Can we add a short delay after dropping the connection?

Applied in https://github.com/ZcashFoundation/zebra/pull/7041/commits/397e70391a5b80a0097b58f740fc1a9eb8c0e7a8.

> If they use different IP addresses (easy with IPv6), they can also add an unlimited number of entries to recent_inbound_connections.

They could add up to `config.peerset_inbound_connection_limit()` entries, then it starts dropping connections without touching `recent_inbound_connections`.

> Do you think it would take a lot of CPU to create and prune twelve thousand entries?

It's a bit better now that it's using `split_off()`, and I don't think it would take 10ms, but I'm not sure.

> Should we put a lower limit on the number of entries?

I think it's okay to rely on `config.peerset_inbound_connection_limit()`.

Reply:
Great, let's merge it and see how it goes?